olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/util/SubprocessConnection.java

```diff
@@ -0,0 +1,296 @@
+            pb.redirectError(ProcessBuilder.Redirect.INHERIT);
+            pb.environment().putAll(environment);
+            process = pb.start();
+            synchronized (process) {
+                BufferedReader stdout = new BufferedReader(new InputStreamReader(process.getInputStream()));
+                String line;
+                while ((line = stdout.readLine()) != null) {
+                    logger.fine("RECEIVED::" + line);
+                    if (line.startsWith("Ready")) {
+                        logger.info("Subprocess is ready");
+                        break;
+                    }
+                }
+                lastIOTime = System.currentTimeMillis();
+            }
+            //
+            // Notify our listeners that we just started
+            listeners.forEach(l -> l.subprocessStarted(this));
+        }
+        //
+        // If the idler isn't running, start it up too. Check if idle time has elapsed every 30 seconds.
+        if (timeoutMillis != 0 && idlerTimer == null) {
+            idlerTimer = new Timer("SubProcessIdler", true);
+            idlerTimer.schedule(new Idler(), 30 * 1000, 30 * 1000);
+        }
+    }
+
+    /**
+     * Shuts down the running subprocess. Attempts to shut down gracefully by issuing the
+     * text "SHUTDOWN" on a line by itself to the process. If the process does not exit
+     * after 5 seconds, the process is terminated forcibly.
+     */
+    public void shutdown() {
+        if (process != null) {
+            synchronized (process) {
+                if (process.isAlive()) {
+                    //
+                    // Announce that we're about to shut down.
+                    listeners.forEach(l -> l.subprocessPreShutdown(this));
+
+                    // First, tell it to shut down, then wait a short
+                    // period before making sure it is down.
+                    PrintWriter stdin = new PrintWriter(process.getOutputStream());
+                    stdin.println(SHUTDOWN);
+                    stdin.flush();
+                    try {
+                        if (!process.waitFor(5, TimeUnit.SECONDS)) {
+                            process.destroyForcibly();
+                        }
+                    } catch (InterruptedException e) {
+                        logger.log(Level.WARNING, "Shutdown interrupted", e);
+                        process.destroyForcibly();
+                    }
+                    //
+                    // Announce that shutdown has happened.
+                    listeners.forEach(l -> l.subprocessPostShutdown(this));
+                }
+            }
+        }
+        if (idlerTimer != null) {
+            idlerTimer.cancel();
+            idlerTimer = null;
+        }
+
+    }
+
+    /**
+     * A thread that will watch for the subprocess to have become idle and
+     * terminate it if the idle timeout period has elapsed.
+     */
+    public class Idler extends TimerTask {
```

Comment:
Maybe this should be final too?

Reply:
As this provides no contract other than that in TimerTask, I think it is safe to allow this to be extended on the off chance somebody wants to change its behavior.